transmission/transmission.go

```diff
@@ -84,57 +86,97 @@ func (h *Honeycomb) Start() error {
 	}
 	h.Logger.Printf("default transmission starting")
 	h.responses = make(chan Response, h.PendingWorkCapacity*2)
-	h.muster.MaxBatchSize = h.MaxBatchSize
-	h.muster.BatchTimeout = h.BatchTimeout
-	h.muster.MaxConcurrentBatches = h.MaxConcurrentBatches
-	h.muster.PendingWorkCapacity = h.PendingWorkCapacity
 	if h.Metrics == nil {
 		h.Metrics = &nullMetrics{}
 	}
-	h.muster.BatchMaker = func() muster.Batch {
-		return &batchAgg{
-			userAgentAddition: h.UserAgentAddition,
-			batches:           map[string][]*Event{},
-			httpClient: &http.Client{
-				Transport: h.Transport,
-				Timeout:   60 * time.Second,
-			},
-			blockOnResponse:       h.BlockOnResponse,
-			responses:             h.responses,
-			metrics:               h.Metrics,
-			disableCompression:    h.DisableGzipCompression || h.DisableCompression,
-			enableMsgpackEncoding: h.EnableMsgpackEncoding,
+	if h.batchMaker == nil {
+		h.batchMaker = func() muster.Batch {
+			return &batchAgg{
+				userAgentAddition: h.UserAgentAddition,
+				batches:           map[string][]*Event{},
+				httpClient: &http.Client{
+					Transport: h.Transport,
+					Timeout:   60 * time.Second,
+				},
+				blockOnResponse:       h.BlockOnResponse,
+				responses:             h.responses,
+				metrics:               h.Metrics,
+				disableCompression:    h.DisableGzipCompression || h.DisableCompression,
+				enableMsgpackEncoding: h.EnableMsgpackEncoding,
+			}
 		}
 	}
+
+	m := h.createMuster()
+	h.muster = m
 	return h.muster.Start()
 }
 
+func (h *Honeycomb) createMuster() *muster.Client {
+	m := new(muster.Client)
+	m.MaxBatchSize = h.MaxBatchSize
+	m.BatchTimeout = h.BatchTimeout
+	m.MaxConcurrentBatches = h.MaxConcurrentBatches
+	m.PendingWorkCapacity = h.PendingWorkCapacity
+	m.BatchMaker = h.batchMaker
+	return m
+}
+
 func (h *Honeycomb) Stop() error {
 	h.Logger.Printf("Honeycomb transmission stopping")
 	err := h.muster.Stop()
 	close(h.responses)
 	return err
 }
 
+func (h *Honeycomb) Flush() (err error) {
+	m := h.muster
+	defer func() {
+		e := m.Stop() // This is latent, so we do not want to execute this with the lock.
+		if err == nil {
+			err = e
+			// bubble up the error from Stop if there was no error from Start
+		}
+	}()
+	h.musterLock.Lock()
+	defer h.musterLock.Unlock()
+	newMuster := h.createMuster()
+	h.muster = newMuster
+	return h.muster.Start()
+}
+
 func (h *Honeycomb) Add(ev *Event) {
+	added := h.tryAdd(ev)
+	if added {
+		h.Metrics.Increment("messages_queued")
+		return
+	}
+	h.Metrics.Increment("queue_overflow")
+	r := Response{
+		Err:      errors.New("queue overflow"),
+		Metadata: ev.Metadata,
+	}
+	h.Logger.Printf("got response code %d, error %s, and body %s",
+		r.StatusCode, r.Err, string(r.Body))
+	writeToResponse(h.responses, r, h.BlockOnResponse)
+}
+
+// tryAdd attempts to add ev to the underlying muster. It returns false if this
+// was unsucessful because the muster queue (muster.Work) is full.
+func (h *Honeycomb) tryAdd(ev *Event) bool {
```

Comment:
I'm a little uncomfortable with the specific error (queue is full) being masked by a boolean - so the error occurs in this function but is named and handled in the calling function. OTOH, creating an error type and returning `error` here seems like overkill for a function that's only used in one place. It's internal, so there are no future implications to changing the signature if somehow it figures out how to handle different error types. (the only one I can think of now is a world in which `Add()` gets a timeout instead of being "block forever" or "don't block at all".) I think in the end I'm fine with the bool, especially since it makes the `if tryAdd(ev)` statement so nice and straightforward. If we ever augment it somehow that'll warrant making this function's ability to communicate what happened more important and we can do it then.